mempool/src/pool.rs

```diff
@@ -0,0 +1,330 @@
+        self.store.contains_outpoint(outpoint) || self.chain_state.contains_outpoint(outpoint)
+    }
+
+    fn validate_transaction(&self, tx: &Transaction) -> Result<(), TxValidationError> {
+        if tx.get_inputs().is_empty() {
+            return Err(TxValidationError::NoInputs);
+        }
+
+        if tx.get_outputs().is_empty() {
+            return Err(TxValidationError::NoOutputs);
+        }
+
+        if tx.is_coinbase() {
+            return Err(TxValidationError::LooseCoinbase);
+        }
+
+        // TODO consier a MAX_MONEY check reminiscent of bitcoin's
+        // TODO consider rejecting non-standard transactions (for some definition of standard)
+        if has_duplicate_entry(tx.get_inputs().iter().map(|input| input.get_outpoint())) {
+            return Err(TxValidationError::DuplicateInputs);
+        }
+
+        if tx.encoded_size() > MAX_BLOCK_SIZE_BYTES {
+            return Err(TxValidationError::ExceedsMaxBlockSize);
+        }
+
+        if self.contains_transaction(&tx.get_id()) {
+            return Err(TxValidationError::TransactionAlreadyInMempool);
+        }
+
+        tx.get_inputs()
+            .iter()
+            .filter_map(|input| self.store.find_conflicting_tx(input.get_outpoint()))
+            .all(|tx| tx.is_replaceable())
+            .then(|| ())
+            .ok_or(TxValidationError::ConflictWithIrreplaceableTransaction)?;
+
+        self.verify_inputs_available(tx)?;
+
+        Ok(())
+    }
+}
+
+impl<C: ChainState + Debug> Mempool<C> for MempoolImpl<C> {
+    fn create(chain_state: C) -> Self {
+        Self {
+            store: MempoolStore::new(),
+            chain_state,
+        }
+    }
+
+    fn new_tip_set(&mut self) -> Result<(), MempoolError> {
+        unimplemented!()
+    }
+    //
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError> {
+        // TODO (1). First, we need to decide on criteria for the Mempool to be considered full. Maybe number
+        // of transactions is not a good enough indicator. Consider checking mempool size as well
+        // TODO (2) What to do when the mempool is full. Instead of rejecting Do incoming transaction we probably want to evict a low-score transaction
+        if self.store.txs_by_fee.len() >= MEMPOOL_MAX_TXS {
+            return Err(MempoolError::MempoolFull);
+        }
+        self.validate_transaction(&tx)?;
+        self.store.add_tx(tx.clone())?;
+        self.chain_state.apply_tx(&tx);
+        Ok(())
+    }
+
+    fn get_all(&self) -> Vec<&Transaction> {
+        self.store.txs_by_fee.iter().map(|mempool_tx| &mempool_tx.tx).collect()
+    }
+
+    fn contains_transaction(&self, tx_id: &Id<Transaction>) -> bool {
+        self.store.txs_by_id.contains_key(&tx_id.get())
+    }
+
+    // TODO Consider returning an error
+    fn drop_transaction(&mut self, tx_id: &Id<Transaction>) {
+        self.store.drop_tx(tx_id);
+        self.chain_state.drop_tx(tx_id)
```

Comment:
I don't understand why dropping a transaction from the mempool would have an effect on the chain_state.

Reply:
It shouldn't that was me being confused. Fixed now.